cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -109,6 +109,39 @@ void ReplaceNullWithOtherType(ValueDescr* first, size_t count) {
   }
 }
 
+void ReplaceTemporalTypes(const TimeUnit::type unit, std::vector<ValueDescr>* descrs) {
+  auto* end = descrs->data() + descrs->size();
+  const std::string* timezone = nullptr;
+
+  for (auto* it = descrs->data(); it != end; it++) {
+    switch (it->type->id()) {
+      case Type::TIMESTAMP: {
+        const auto& ty = checked_cast<const TimestampType&>(*it->type);
+        timezone = &ty.timezone();
+        if (timezone) {
+          it->type = timestamp(unit, *timezone);
+        } else {
+          it->type = timestamp(unit);
+        }
```

Comment:
Can't this just be `it->type = timestamp(unit, ty.timezone())`? Since `timezone()` is empty string for naive timestamps (not nullptr)

Reply:
Will do. (I was trying to get around `Bad dynamic_cast`.)